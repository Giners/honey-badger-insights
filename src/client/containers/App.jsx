import 'isomorphic-fetch'
import { ApolloClient } from 'apollo-client'
import { HttpLink } from 'apollo-link-http'
import { InMemoryCache } from 'apollo-cache-inmemory'

import React from 'react'

import HoneyBadgersTable from './../components/HoneyBadgersTable'
import {
  autonomousSystemsQuery,
  geoLocationsQuery,
  topHoneyBadgersQuery,
} from './../graphql/queries'
import config from './../../server/config'

const apolloClient = new ApolloClient({
  link: new HttpLink({
    uri: `${config.app.hostURL}:${config.app.port}${
      config.serviceURIs.graphQL
    }`,
  }),
  cache: new InMemoryCache(),
})

/** The initial state of the App component when it is mounted. Exported for testing purposes. */
export const initialState = { honeyBadgers: new Map(), error: null }

/**
 * The <App> component is a stateful React component that is responsible for providing the "root"
 * for our visual and logical React components that make up our app. Its most important job is to
 * query for information about honey badgers and then to pass that information along to child
 * components so they can render it into a meaningful visualization.
 */
class App extends React.Component {
  /**
   * Helper method that makes a query to the GraphQL service/API endpoint for the top honey badgers
   * that have been seen in the past 24 hours.
   *
   * @return {Promise<Map<string, HoneyBadger>, error>} Returns a promise where it is resolved to a
   * map that contains info about the top honey badgers. The map is keyed off of the honey badgers
   * IP address. Should this fail to query the GraphQL service/API endpoint or for any other reason
   * this promise will be rejected with the error.
   */
  static async collectTopHoneyBadgersInfo() {
    const { topHoneyBadgers } = (await apolloClient.query({
      query: topHoneyBadgersQuery,
    })).data

    const thbKVPs = topHoneyBadgers.map(topHoneyBadger => [
      topHoneyBadger.ipAddress,
      topHoneyBadger,
    ])
    return new Map(thbKVPs)
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about the
   * autonomous systems that is responsible/in charge of said honey badger.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the autonomous systems that has been
   * collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about autonomous systems as well as add the info to the map of
   * honey badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectAutonomousSystemsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { autonomousSystems } = (await apolloClient.query({
      query: autonomousSystemsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the autonomous system that the honey badger falls under with the
    // existing data about honey badgers
    autonomousSystems.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, as: { ...rest } })
    })
  }

  /**
   * Helper method that given a map of honey badgers where they are keyed off of their IP address
   * will make a query to the GraphQL service/API endpoint for said honey badger for info about
   * where said honey badger is located geospatially.
   *
   * Note that this function isn't pure and has side effects in that it modifies the map of honey
   * badgers passed into it to add in the info about the locations that has been collected.
   *
   * @param {Map<string, HoneyBadger>} honeyBadgers - Map of honey badgers where they are keyed off
   * of their IP address
   *
   * @return {Promise<undefined, error>} Returns a promise where it will resolve if able to
   * successfully query for info about the locations as well as add the info to the map of honey
   * badgers that is passed into it. Should we fail to query the GraphQL service/API endpoint
   * or for any other reason this promise will be rejected with the error.
   */
  static async collectGeoLocationsInfo(honeyBadgers) {
    const ipAddresses = [...honeyBadgers.keys()]

    const { geoLocations } = (await apolloClient.query({
      query: geoLocationsQuery,
      variables: { ipAddresses },
    })).data

    // Join the data about the location of the honey badger with the existing data about honey
    // badgers
    geoLocations.forEach(({ ipAddress, ...rest }) => {
      const honeyBadger = honeyBadgers.get(ipAddress)

      honeyBadgers.set(ipAddress, { ...honeyBadger, geoLocation: { ...rest } })
    })
  }

  /**
   * Creates a new stateful React component that is responsible for providing the "root" for the
   * visual and logical React components that make up the app.
   */
  constructor() {
    super()

    this.state = initialState
  }

  /**
   * React component lifecycle method that will be called after the <App> component has been mounted
   * into the DOM. After which we make calls to our GraphQL service/API endpoint to collect
   * information about honey badgers and update our state with it.
   */

  // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
  /* eslint-disable react/no-did-mount-set-state */
  async componentDidMount() {
    try {
      // Query for the top honey badgers from our GraphQL service/API endpoint and enqueue a request
      // to set the apps state to the honey badgers
      const honeyBadgers = await App.collectTopHoneyBadgersInfo()

      this.setState({ honeyBadgers })

      // Now collect the following info about honey badgers from our GraphQL service/API endpoint
      // and enqueue a request to set the state of the app with the new info we collected:
      // * The autonomous systems that oversee the IP address of the honey badger
      // * The geospatial location of a honey badgers IP address
      await App.collectAutonomousSystemsInfo(honeyBadgers)
      this.setState({ honeyBadgers })

      await App.collectGeoLocationsInfo(honeyBadgers)
      this.setState({ honeyBadgers })
    } catch (error) {
      // It is okay to call 'setState()' in 'componentDidMount()' as discussed in the React docs
      // eslint-disable-next-line react/no-did-mount-set-state
      this.setState({ error })
    }
  }

  render() {
    const { honeyBadgers, error } = this.state

    if (error) {
      return `The following error occurred: ${error}`
    }

    return <HoneyBadgersTable honeyBadgers={[...honeyBadgers.values()]} />
  }
}

export default App
