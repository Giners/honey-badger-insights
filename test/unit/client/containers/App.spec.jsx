/* eslint prefer-arrow-callback: 0, func-names: 0 */
/* eslint no-unused-expressions: 0, chai-friendly/no-unused-expressions: 2 */

// General 3rd-party supporting test libs
import chai, { expect } from 'chai'
import chaiJestSnapshot from 'chai-jest-snapshot'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

// General 3rd-party supporting libs
import React from 'react'
import List, { ListItem, ListItemText } from 'material-ui/List'

// App React component under test
import App, { initialState } from './../../../../src/client/containers/App'

// Configure Chai to work with Jest
chai.use(chaiJestSnapshot)

describe('React component test: <App>', function() {
  beforeEach(
    'Configure chai-jest-snapshot for "Mocha configuration mode"',
    function() {
      chaiJestSnapshot.configureUsingMochaContext(this)
      chaiJestSnapshot.setFilename(`${__filename}.snap`)
    },
  )

  describe('Renders correctly for given application state:', function() {
    // Reference to an Enzyme shallow wrapper around our <App> component that can be used throughout
    // the following tests
    let appWrapper = null

    beforeEach('Setup Enzyme wrapper', function() {
      appWrapper = shallow(<App />)

      expect(appWrapper).to.not.be.null
      expect(appWrapper.exists()).to.be.true
    })

    it('Initial state', function() {
      expect(appWrapper.state()).to.equal(initialState)

      // We expect a list to be rendered without any elements in it since we shouldn't have any
      // honey badgers when we first mount
      expect(appWrapper.find(List).exists()).to.be.true
      expect(appWrapper.find(ListItem).exists()).to.be.false

      expect(toJson(appWrapper)).to.matchSnapshot()
    })

    it('Error (state contains empty list of honey badgers)', function() {
      const error = 'Bananas LOL'
      appWrapper.setState({ error })

      // If an error is set then we ought to render just the text error
      expect(appWrapper.text()).to.include(error)

      expect(toJson(appWrapper)).to.matchSnapshot()
    })

    it('Error (state contains non-empty list of honey badgers)', function() {
      const error = 'Bananas LOL'
      appWrapper.setState({ error, honeyBadgers: [{}] })

      // Even if we have some honey badgers in our list we ought to still render an error if one is
      // provided
      expect(appWrapper.text()).to.include(error)

      expect(toJson(appWrapper)).to.matchSnapshot()
    })

    it('Honey badgers (state contains non-empty list of honey badgers)', function() {
      appWrapper.setState({ honeyBadgers: [{}] })

      // We expect a list to be rendered with elements if we have any honey badgers to render
      expect(appWrapper.find(List).exists()).to.be.true
      expect(appWrapper.find(ListItemText).exists()).to.be.true

      expect(toJson(appWrapper)).to.matchSnapshot()
    })
  })
})
