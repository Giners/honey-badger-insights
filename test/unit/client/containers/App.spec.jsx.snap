// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React component test: <App> Renders correctly for given application state: Error (state contains empty list of honey badgers) 1`] = `"The following error occurred: Bananas LOL"`;

exports[`React component test: <App> Renders correctly for given application state: Error (state contains non-empty list of honey badgers) 1`] = `"The following error occurred: Bananas LOL"`;

exports[`React component test: <App> Renders correctly for given application state: Honey badgers (state contains non-empty list of honey badgers) 1`] = `
<div>
  <WithStyles(Typography)
    align="center"
    variant="display3"
  >
    Honey Badger Insights
  </WithStyles(Typography)>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Tabs)
      centered={true}
      onChange={[Function]}
      value={0}
    >
      <WithStyles(Tab)
        label="Honey Badgers Data Table"
      />
      <WithStyles(Tab)
        label="About Honey Badgers Insights"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <HoneyBadgersTable
    honeyBadgers={
      Array [
        Object {},
      ]
    }
  />
</div>
`;

exports[`React component test: <App> Renders correctly for given application state: Initial props/state 1`] = `
<div>
  <WithStyles(Typography)
    align="center"
    variant="display3"
  >
    Honey Badger Insights
  </WithStyles(Typography)>
  <WithStyles(AppBar)
    position="static"
  >
    <WithStyles(Tabs)
      centered={true}
      onChange={[Function]}
      value={0}
    >
      <WithStyles(Tab)
        label="Honey Badgers Data Table"
      />
      <WithStyles(Tab)
        label="About Honey Badgers Insights"
      />
    </WithStyles(Tabs)>
  </WithStyles(AppBar)>
  <HoneyBadgersTable
    honeyBadgers={Array []}
  />
</div>
`;
